### ISSUE TITLE
Not sure of status of Async specs for Scalajs

### ISSUE THREAD

**Issue by benhutchison** [etorreborre/specs2#717]
I raised a discussion on github around [testing code with `Future`s on Scalajs](https://gitter.im/etorreborre/specs2?at=5bc927cfc08b8b306737ee75), and subsequently @bpholt raised [similar needs for `IO`](https://gitter.im/etorreborre/specs2?at=5bd9f3d40445e15fad01a1d7).

Summary is that it's not clearly documented now if async tests work on Scalajs. My own experiment ran into problems. I couldnt find unit tests running on scalajs for async mode. So it would be good to clarify whether its considered to work and maybe needs an example, or if it isn't known to work and needs some dev.

**Comment by etorreborre** [comment]
I've added the possibility to return `Future[R]` (where `R : AsResult`) in an example on master. I have published a version where you can use futures in a mutable spec if you want to try: `4.3.5-78abffa2e-20181150936`. The drawback is that you cannot set retries or timeouts specifically for a given example. 

I still need to make sure this will all work with immutable specs as well.

**Comment by benhutchison** [comment]
From a first try, seems to be working for me in scalajs :)

One problem I hit. I couldn't pass an implicit EC like on [your example](https://github.com/etorreborre/specs2/commit/78abffa2e8fa312143eaf14441adf46484c1b093#diff-d4b502e58184da449c29446b9e95367cR8), it seemed like the Spec's constructor needed to be no-arg, as I got this:
```
java.lang.InstantiationError: scalajswebsocket.EchoServiceIntegrationSpec
  at $c_jl_InstantiationError.$c_jl_Throwable.fillInStackTrace__jl_Throwable(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:41361:14)
  at $c_jl_InstantiationError.$c_jl_Throwable.init___T__jl_Throwable__Z__Z(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:41473:10)
  at java.lang.InstantiationError.<init>(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:88169:58)
  at org.scalajs.testinterface.TestUtils$.newInstance(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:14302:44)
  at {anonymous}()(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:86176:63)
  at scala.scalajs.runtime.AnonFunction0.apply(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:55669:23)
  at org.specs2.fp.Need$$anon$2.value0$lzycompute(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:49737:32)
  at org.specs2.fp.Need$$anon$2.value0(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:49730:37)
  at org.specs2.fp.Need$$anon$2.value(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:49733:15)
  at {anonymous}()(file:///Users/ben/workspace/scalajs-websocket/target/scala-2.12/scalajs-websocket-test-fastopt.js:35668:23)
```

There's a EC provided at `scala.scalajs.concurrent.JSExecutionContext.queue` for scalajs only specs, which is what Im writing ATM, so I used that successfully. However, in future there will surely be need the need for async cross-JVM/JS specs, and Im not sure yet what EC they are meant to run on that's suitable for both environments..

**Comment by etorreborre** [comment]
I think you can use the global EC for ScalaJS specifications: https://stackoverflow.com/questions/32437585/cannot-find-an-implicit-executioncontext-error-in-scala-js-example-app. This should also work for cross JVM/JS specifications since specs2 has its own execution environment in the JVM case.

**Comment by benhutchison** [comment]
There is an Instantiation Error in there. I don't think your specs can have
implicit as ctor args in scalajs. Not sure why but removing it should work.

**Comment by benhutchison** [comment]
I think this is where the `Nil` list of `paramTypes` is coming from: https://github.com/etorreborre/specs2/blob/master/common/js/src/main/scala/org/specs2/reflect/Classes.scala#L16

**Comment by benhutchison** [comment]
Related:
https://github.com/scala-js/scala-js/issues/2733
https://github.com/sjrd/scalajs-reflect

Getting the constructor signature of the Test class doesn't seem easy ATM, which is needed to decide if 0- or 1- param; it's not available from its `ClassTag` AFAIK. 

There seems a workaround - don't take an implicit EC and just use the global one, in specs that need be run across JS/JVM.

**Comment by etorreborre** [comment]
> There seems a workaround - don't take an implicit EC and just use the global one, in specs that need be run across JS/JVM.

Yes I think that's the way to go for now. This is hopefully won't be a big deal

**Comment by benhutchison** [comment]
Ok with this decided, I think this issue can be closed and a build published, my experience is that the async tests worked fine

**Comment by benhutchison** [comment]
The support for async Scalajs tests is currently only in `"specs2-core" % "4.3.5-78abffa2e-20181150936"`, please publish a 4.3.6 version if this is a now supported feature.

**Comment by etorreborre** [comment]
Sure, I'm really behind on my publishing... I'll do that today

**Comment by bpholt** [comment]
I'm working on #724 and I was trying to build some implicit cats-effect instances based on the execution context in the specs2 `ExecutionEnv`, and I ran into this issue on the JS side. I got it working with this workaround, though:

```scala
  def createInstance[T <: AnyRef](className: String, loader: ClassLoader, defaultInstances: =>List[AnyRef] = Nil)(implicit m: ClassTag[T]): Operation[T] = {
    def newInstanceWithOptionalExecutionEnvironment() =
      try {
        TestUtils.newInstance(className, loader, Nil)(Nil).asInstanceOf[T]
      } catch {
        case _: InstantiationError =>
          TestUtils.newInstance(className, loader, Seq(classOf[ExecutionEnv]))(List(defaultInstances.tail.head)).asInstanceOf[T]
      }

    if (className.endsWith("$")) protect[OperationStack, T](TestUtils.loadModule(className.dropRight(1), loader).asInstanceOf[T])
    else                         protect[OperationStack, T](newInstanceWithOptionalExecutionEnvironment())
  }
```

AFAICT this lets me require `(implicit ee: ExecutionEnv)` in the constructor of both JVM and JS tests.

**Comment by symbiont-eric-torreborre** [comment]
@bpholt I just saw this issue again. It is fixed now in [`4.10.0`](https://github.com/etorreborre/specs2/releases/tag/SPECS2-4.10.0)
